config.org
#+TITLE: Marty Buchaus Config
#+startup: content


* Config
** File Header
*** Top

#+begin_src emacs-lisp
;; $doomdir/config.el --- summary -*- lexical-binding: t; no-byte-compile: t; -*-
;;
;; author: marty buchaus <marty@dabuke.com>
;; copyright © 2021, marty buchaus, all rights reserved.
;; created:  1 November 2021
;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
#+end_src

*** Notes

#+begin_src emacs-lisp

;;; Notes
;;;; Global

;;  2021 11 18 Update clean Install and config
;;  2021 10 12  added code from Stuff from  https://github.com/Artawower/.doom/blob/main/config.el#L308

#+end_src

** Global

#+begin_src emacs-lisp
;;; Global

(setq user-full-name "Marty Buchaus")
(setq user-mail-address "marty@dabuke.com")
(setq epg-gpg-program "/usr/bin/gpg2")

(setq-default enable-local-variables t)            ; allow for reading the local variables file
(setq-default window-combination-resize t)
(setq-default left-margin-width 1)                 ; Define new widths
(setq-default right-margin-width 2)                ; Define new widths.
(setq-default x-stretch-cursor t)

(setq scroll-margin 2)                             ; it's nice to maintain a little margin
(setq evil-want-fine-undo t)                       ; by default while in insert all changes are one big blob. be more granular
(setq undo-limit 80000000)                         ; raise undo-limit to 80mb
(setq auto-save-default t)                         ; nobody likes to loose work, i certainly don't
(setq confirm-kill-emacs nil)                      ; stop hounding me and quit
(setq display-time-24hr-format t)                  ; I wonder what this does
(setq password-cache-expiry nil)                   ; i can trust my computers ... can't i?
(setq read-process-output-max (* 1024 1024))
(setq truncate-string-ellipsis "…")                ; unicode ellispis are nicer than "...", and also save /precious/ space
(setq warning-minimum-level :emergency)
(setq garbage-collection-messages t)

(setq doom-scratch-initial-major-mode 'lisp-interaction-mode)  ; Make the scratch buffer start in lisp mode

(display-time-mode 1)                              ; enable time in the mode-line

(global-subword-mode 1)                            ; CamelCase and it makes refactoring slightly Essie

(after! projectile
  (setq projectile-project-search-path '("~/Source")))
;; this doesn't seem to work to fix the doom doctor issue
;; (setq projectile-project-root-files-bottom-up (remove ".git" projectile-project-root-files-bottom-up)))

(set-window-buffer nil (current-buffer))
(setenv "zstd" "/usr/bin/zstd")

#+end_src

** Vtirm Settings

#+begin_src emacs-lisp

;;;;; VTERM

(setq vterm-kill-buffer-on-exit t)
(setq vterm-always-compile-module t)               ; Always compile the vterm module
(setq vterm-shell "/usr/bin/zsh")

#+end_src

** UI
*** Fonts
#+begin_src emacs-lisp

;;; Set Fonts
(setq doom-font (font-spec :family "DejaVu Sans Mono" :size 15 :weight 'regular )
      doom-variable-pitch-font (font-spec :family "Ubuntu" :style "Regular" :size 15 :weight 'regular)
      doom-unicode-font (font-spec :family "symbola" :size 14)
      doom-big-font (font-spec :family "DejaVu Sans Mono" :size 24))

#+end_src

*** Faces
#+begin_src emacs-lisp
;;;;; FACES

(custom-set-faces!
  '(font-lock-comment-face :slant italic)
  '(font-lock-keyword-face :slant italic))
(setq global-prettify-symbols-mode t)

(custom-set-faces!
 '(mode-line :family "firacode nerd font mono" :height 100)
 '(mode-line-inactive :family "firacode nerd font mono" :height 100))

#+end_src

*** Theme
#+begin_src emacs-lisp

;;;;; THEME

(setq doom-theme 'doom-dracula )

(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))

#+end_src
*** Modeline
#+begin_src emacs-lisp

;;;;; MODELINE

(after! doom-modeline
  (setq all-the-icons-scale-factor 1.1)
  (setq doom-modeline-enable-word-count t)
  (setq doom-themes-padded-modeline t)
  (setq auto-revert-check-vc-info t)
  (setq doom-modeline-buffer-file-name-style 'relative-to-project)
  (setq doom-modeline-major-mode-color-icon (display-graphic-p))
  (setq doom-modeline-major-mode-icon (display-graphic-p))
  (setq doom-modeline-vcs-max-length 60)

  (add-hook! 'doom-modeline-mode-hook
    (progn
      (set-face-attribute 'header-line nil
                          :background (face-background 'mode-line)
                          :foreground (face-foreground 'mode-line))
      )))
#+end_src
*** Line Numbers
#+begin_src emacs-lisp

;;;;; LINE NUMBERS

(setq display-line-numbers-type 'relative)

;; remove numbers from these modes

(dolist (mode '(org-mode-hook
                term-mode-hook
                shell-mode-hook
                eshell-mode-hook))
  (add-hook mode (lambda () (display-line-numbers-mode 0))))

#+end_src

** Packages
*** Spelling
#+begin_src emacs-lisp

;;;; SPELLING

(use-package! flyspell
  :defer t
  :config
  ;; (setq ispell-program-name "aspell")
  ;; You could add extra option "--camel-case" for since Aspell 0.60.8
  ;; @see https://github.com/redguardtoo/emacs.d/issues/796
  ;; (setq ispell-extra-args '("--sug-mode=ultra" "--lang=en_US" "--run-together" "--run-together-limit=16"))
  (setq-default flyspell-prog-text-faces
                '(tree-sitter-hl-face:comment
                  tree-sitter-hl-face:doc
                  tree-sitter-hl-face:string
                  tree-sitter-hl-face:function
                  tree-sitter-hl-face:variable
                  tree-sitter-hl-face:type
                  tree-sitter-hl-face:method
                  tree-sitter-hl-face:function.method
                  tree-sitter-hl-face:function.special
                  tree-sitter-hl-face:attribute
                  font-lock-comment-face
                  font-lock-doc-face
                  font-lock-string-face
                  lsp-face-highlight-textual
                  default))


  (setq flyspell-lazy-idle-seconds 2)
  (setq ispell-personal-dictionary "~/.config/doom/dictionary/ispell_personal" )
  (setq spell-fu-directory "~/.config/doom/dictionary") ;; Please create this directory manually.
  (after! ispell
    (setq ispell-program-name "aspell"
          ;; Notice the lack of "--run-together"
          ispell-extra-args '("--sug-mode=ultra" "--lang=en_US" "--run-together" "--run-together-limit=16"))
    (ispell-kill-ispell t))

  (defun flyspell-buffer-after-pdict-save (&rest _)
    (flyspell-buffer))

  (advice-add 'ispell-pdict-save :after #'flyspell-buffer-after-pdict-save))


(add-hook 'text-mode-hook 'flyspell-mode!)
(add-hook 'prog-mode-hook 'flyspell-prog-mode)
#+end_src
*** DIRED
#+begin_src emacs-lisp

;;;; DIRED

(add-hook 'peep-dired-hook 'evil-normalize-keymaps)
;; get file icons in dired
(add-hook 'dired-mode-hook 'all-the-icons-dired-mode)
;; with dired-open plugin, you can launch external programs for certain extensions
;; for example, i set all .png files to open in 'sxiv' and all .mp4 files to open in 'mpv'
(setq dired-open-extensions '(("gif" . "sxiv")
                              ("jpg" . "sxiv")
                              ("png" . "sxiv")
                              ("mkv" . "mpv")
                              ("mp4" . "mpv")))
#+end_src
*** Company
#+begin_src emacs-lisp

;;;; COMPANY

(setq company-idle-delay 0.5)
#+end_src

*** Dashboard

#+begin_src emacs-lisp

;;;; DASHBOARD

(use-package! dashboard
	      :init      ;; tweak dashboard config before loading it
	      (setq dashboard-set-heading-icons t)
	      (setq dashboard-set-file-icons t)
	      (setq dashboard-banner-logo-title "emacs is more than a text editor!")
	      ;;(setq dashboard-startup-banner 'logo) ;; use standard emacs logo as banner
	      (setq dashboard-startup-banner "~/.config/doom/banners/doom-emacs-slant-out-bw.png")  ;; use custom image as banner
	      (setq dashboard-center-content t) ;; set to 't' for centered content
	      (setq dashboard-items '((recents . 5)
				      (agenda . 5 )
				      (bookmarks . 5)
				      (projects . 5)
				      (registers . 5)))
	      :config
	      (dashboard-setup-startup-hook)
	      (dashboard-modify-heading-icons '((recents . "file-text")
						(bookmarks . "book"))))

#+end_src
*** Magit
#+begin_src emacs-lisp

(use-package! magit
	      :config
	      (define-key transient-map        "q" 'transient-quit-one)
	      (define-key transient-edit-map   "q" 'transient-quit-one)
	      (define-key transient-sticky-map "q" 'transient-quit-seq)
	      (setq magit-revision-show-gravatars '("^author:     " . "^commit:     ")))

#+end_src
** ORG-Mode
*** Pre org config

Set the Org and Roam Directories so that subsequent runs will have this variable
Set also the org-agenda files variable

#+begin_src emacs-lisp

(setq org-directory "~/Nextcloud/Notes/org/")
(setq org-roam-directory "~/Nextcloud/Notes/org/")
(setq org-roam-dailies/directory "daily/")
(setq org-contacts-files '("~/Nextcloud/Notes/org/contacts.org"))

(setq  marty/org-agenda-files (list
                               (concat org-directory "Tasks.org")
                               (concat org-directory "Habits.org")
                               (concat org-directory "Calendar.org")
                               (concat org-directory "contacts.org")
                               (concat org-directory "Someday.org")
                               (concat org-directory "0mobile.org")))

#+end_src
*** Org Publish

#+begin_src emacs-lisp

;;;;; PUBLISH ALIST

(after! org
  (defun marty/publish (a b c)
    (setq org-export-with-toc t)
    (org-html-publish-to-html a b c))

  (require 'find-lisp)
  (defun marty/publish-NSI-Documentation (a b c)
    (setq org-export-with-toc t)
    (let ((org-id-extra-files (find-lisp-find-files "~/Source/NSI/NSI-Documentation/" "\.org$")))
      (org-html-publish-to-html a b c)))

  (setq org-publish-project-alist
        '(
          ("NSI-Documentation-content"
           :base-directory "~/Source/NSI/NSI-Documentation/"
           :base-extension "org"
           :publishing-directory "~/Source/NSI/NSI-Documentation/docs"
           :publishing-function marty/publish-NSI-Documentation
           :exclude "Archive"
           :section-numbers nil
           :with-toc nil
           :auto-sitemap t
           :sitemap-filename "filemap.org"
           :sitemap-title "& Sitemap"
           :headline-levels 10
           :auto-preamble t
           :recursive t)

          ("NSI-Documentation-images"
           :base-directory "~/Source/NSI/NSI-Documentation/images/"
           :base-extension "jpg\\|gif\\|png"
           :publishing-directory "~/Source/NSI/NSI-Documentation/docs/images/"
           :publishing-function org-publish-attachment
           :recursive t)

          ("NSI-Documentation-TVA-ScanReports-2020-images"
           :base-directory "~/Source/NSI/NSI-Documentation/TVA/ScanReports/2020/images/"
           :base-extension "jpg\\|gif\\|png"
           :publishing-directory "~/Source/NSI/NSI-Documentation/docs/TVA/ScanReports/2020/images/"
           :publishing-function org-publish-attachment
           :recursive t)

          ("NSI-Documentation-TVA-ScanReports-2020-reports"
           :base-directory "~/Source/NSI/NSI-Documentation/TVA/ScanReports/2020/reports/"
           :base-extension "ods\\|csv\\|xls\\|xslt\\|pdf"
           :publishing-directory "~/Source/NSI/NSI-Documentation/docs/TVA/ScanReports/2020/reports/"
           :publishing-function org-publish-attachment
           :recursive t)

          ("NSI-Documentation-TVA-ScanReports-2021-images"
           :base-directory "~/Source/NSI/NSI-Documentation/TVA/ScanReports/2021/images/"
           :base-extension "jpg\\|gif\\|png"
           :publishing-directory "~/Source/NSI/NSI-Documentation/docs/TVA/ScanReports/2021/images/"
           :publishing-function org-publish-attachment
           :recursive t)

          ("NSI-Documentation-TVA-ScanReports-2021-reports"
           :base-directory "~/Source/NSI/NSI-Documentation/TVA/ScanReports/2021/reports/"
           :base-extension "ods\\|csv\\|xls\\|xslt\\|pdf"
           :publishing-directory "~/Source/NSI/NSI-Documentation/docs/TVA/ScanReports/2021/reports/"
           :publishing-function org-publish-attachment
           :recursive t)

          ("NSI-Documentation-TVA-ScanReports-files"
           :base-directory "~/Source/NSI/NSI-Documentation/TVA/ScanReports/files/"
           :base-extension "ods\\|csv\\|xls\\|xslt\\|pdf"
           :publishing-directory "~/Source/NSI/NSI-Documentation/docs/TVA/ScanReports/files/"
           :publishing-function org-publish-attachment
           :recursive t)

          ("salt-master"
           :base-directory "~/Source/NSI/salt-master/"
           :base-extension "org"
           :publishing-directory "~/Source/NSI/salt-master/docs"
           :publishing-function marty/publish
           :exclude "docs"
           :section-numbers nil
           :with-toc nil
           :auto-sitemap t
           :sitemap-filename "filemap.org"
           :sitemap-title "& Sitemap"
           :headline-levels 7
           :auto-preamble t
           :recursive t)

          ("NSI-Documentation" :components ("NSI-Documentation-content"
                                            "NSI-Documentation-images"
                                            "NSI-Documentation-TVA-ScanReports-2020-images"
                                            "NSI-Documentation-TVA-ScanReports-2020-reports"
                                            "NSI-Documentation-TVA-ScanReports-2021-images"
                                            "NSI-Documentation-TVA-ScanReports-2021-reports"
                                            "NSI-Documentation-TVA-ScanReports-files")))))


#+end_src

*** Org-Agenda
Setup and config for the org agenda system of org-mode

#+begin_src emacs-lisp

(after! org-agenda

  (setq org-agenda-block-separator nil)
  (setq org-agenda-compact-blocks t)
  (setq org-agenda-files marty/org-agenda-files)
  (setq org-agenda-include-deadlines t)
  (setq org-agenda-start-on-weekday 1)
  (setq org-agenda-start-with-log-mode t)
  (setq org-agenda-tags-column 100) ;; from testing this seems to be a good value
  (setq org-agenda-window-setup 'current-window)
  (setq org-deadline-warning-days 14)

  ;; Ignore scheduled tasks in task list
  (setq org-agenda-todo-ignore-scheduled 'all)
  (setq org-agenda-todo-ignore-deadlines 'far)

  ;; Skip Finished Items
  (setq org-agenda-skip-deadline-if-done t)
  (setq org-agenda-skip-scheduled-if-done t)

  (require 'org-projectile)
  (mapcar #'(lambda (file)
              (when (file-exists-p file)
                (push file org-agenda-files)))
          (org-projectile-todo-files)))

#+end_src

*** ORG Configuration
**** BASE
#+begin_src emacs-lisp
(after! org

;;;;;; BASE

  (setq org-default-notes-file (concat org-directory "0mobile.org"))
  (setq org-download-image-dir "~/Nextcloud/Notes/images/")
  (setq org-id-locations-file "~/Nextcloud/Notes/org-id-locations")
  (setq org-persp-startup-org-file "~/Nextcloud/Notes/org/0mobile.org")
  (setq org-projectile-file "todo.org")
  (setq org-fancy-priorities-list '("🅰" "🅱" "🅲" "🅳" "🅴"))
  (setq org-clock-sound "~/Nextcloud/Music/sounds/shipsBell.wav")
  (setq org-startup-with-inline-images t)  ; Show Inline Images

;;;;;; CLOCKING

  ;; (setq org-clock-into-drawer "CLOCKING")
  ;; Where to put the clock in and out for tracked items
  (setq org-clock-out-remove-zero-time-clocks t)

;;;;;; LOGGING AND ID

  (setq org-log-done t)
  (setq org-log-into-drawer t)
  (setq org-icalendar-store-UID t)
  (setq org-id-track-globally t)

;;;;;; REFILE TARGETS

  (setq org-refile-targets '((nil :maxlevel . 3)
                             (org-agenda-files :maxlevel . 5)))

  (setq org-refile-use-outline-path 'file)
  (setq org-outline-path-complete-in-steps nil)
  (setq org-refile-allow-creating-parent-nodes 'confirm)

  (setq prettify-symbols-unprettify-at-point 'right-edge)
)
#+end_src

**** Org-Symbols / org-mode-hook

#+begin_src emacs-lisp
;;;;;; org-mode-hook
(after! org
  (add-hook 'org-mode-hook (lambda ()
                             "Beautify Org Checkbox Symbol"
                             (push '("#+ACTIVE:"            . ""  ) prettify-symbols-alist)
                             (push '("#+BEGIN_EXAMPLE"      . "↦"  ) prettify-symbols-alist)
                             (push '("#+BEGIN_HIDDEN"       . ""  ) prettify-symbols-alist)
                             (push '("#+BEGIN_QUOTE"        . "↦"  ) prettify-symbols-alist)
                             (push '("#+BEGIN_SRC"          . "↦"  ) prettify-symbols-alist)
                             (push '("#+CATEGORY:"          . "⛏ "  ) prettify-symbols-alist)
                             (push '("#+CLOSE_SPOILER"      . ""  ) prettify-symbols-alist)
                             (push '("#+END_EXAMPLE"        . "⇤"  ) prettify-symbols-alist)
                             (push '("#+END_HIDDEN"         . ""  ) prettify-symbols-alist)
                             (push '("#+END_QUOTE"          . "⇤"  ) prettify-symbols-alist)
                             (push '("#+END_SRC"            . "⇤"  ) prettify-symbols-alist)
                             (push '("#+FILETAGS:"          . ""  ) prettify-symbols-alist)
                             (push '("#+ID:"                . ""  ) prettify-symbols-alist)
                             (push '("#+STARTUP:"           . "🌟"  ) prettify-symbols-alist)
                             (push '("#+START_SPOILER"      . ""  ) prettify-symbols-alist)
                             (push '("#+TITLE:"             . ""  ) prettify-symbols-alist)
                             (push '("#+begin_example"      . "↦"  ) prettify-symbols-alist)
                             (push '("#+begin_quote"        . "❝"  ) prettify-symbols-alist)
                             (push '("#+begin_src"          . "↦"  ) prettify-symbols-alist)
                             (push '("#+category:"          . "⛏ "  ) prettify-symbols-alist)
                             (push '("#+end_example"        . "⇤"  ) prettify-symbols-alist)
                             (push '("#+end_quote"          . "❞"  ) prettify-symbols-alist)
                             (push '("#+end_src"            . "⇤"  ) prettify-symbols-alist)
                             (push '("#+filetags:"          . ""  ) prettify-symbols-alist)
                             (push '("#+startup:"           . "⏻"  ) prettify-symbols-alist)
                             (push '("#+title:"             . ""  ) prettify-symbols-alist)
                             (push '("---"                  . "—"  ) prettify-symbols-alist)
                             (push '("->"                   . "→"  ) prettify-symbols-alist)
                             (push '("..."                  . "…"  ) prettify-symbols-alist)
                             (push '("::"                   . "∷"  ) prettify-symbols-alist)
                             (push '(":CREATION_TIME:"      . ""  ) prettify-symbols-alist)
                             (push '(":ID:"                 . ""  ) prettify-symbols-alist)
                             (push '(":LAT-LONG:"           . ""  ) prettify-symbols-alist)
                             (push '(":MAIL:"               . ""  ) prettify-symbols-alist)
                             (push '(":attr_html"           . "🄗"  ) prettify-symbols-alist)
                             (push '(":attr_latex"          . "🄛"  ) prettify-symbols-alist)
                             (push '(":attr_org"            . "⒪"  ) prettify-symbols-alist)
                             (push '(":author"              . "𝘼"  ) prettify-symbols-alist)
                             (push '(":beamer_header"       . "🅑"  ) prettify-symbols-alist)
                             (push '(":begin_export"        . "⏩"  ) prettify-symbols-alist)
                             (push '(":caption"             . "☰"  ) prettify-symbols-alist)
                             (push '(":date"                . "𝘿"  ) prettify-symbols-alist)
                             (push '(":end"                 . "∎" ) prettify-symbols-alist)
                             (push '(":end_export"          . "⏪"  ) prettify-symbols-alist)
                             (push '(":header"              . "›"  ) prettify-symbols-alist)
                             (push '(":html"                . "🅗"  ) prettify-symbols-alist)
                             (push '(":html_head"           . "🅷" ) prettify-symbols-alist)
                             (push '(":latex"               . "🅛" ) prettify-symbols-alist)
                             (push '(":latex_class"         . "🄻" ) prettify-symbols-alist)
                             (push '(":latex_header"        . "🅻" ) prettify-symbols-alist)
                             (push '(":macro"               . "𝓜" ) prettify-symbols-alist)
                             (push '(":options"             . "⌥" ) prettify-symbols-alist)
                             (push '(":results"             . "🠶" ) prettify-symbols-alist)
                             (push '("<-"                   . "←" ) prettify-symbols-alist)
                             (push '("[ ]"                  . "☐"  ) prettify-symbols-alist)
                             (push '("[#A]"                 . "⚑"  ) prettify-symbols-alist)
                             (push '("[#B]"                 . "⬆"  ) prettify-symbols-alist)
                             (push '("[#C]"                 . "■"  ) prettify-symbols-alist)
                             (push '("[#D]"                 . "⬇"  ) prettify-symbols-alist)
                             (push '("[#E]"                 . "❓"  ) prettify-symbols-alist)
                             (push '("[-]"                  . "◼" ) prettify-symbols-alist)
                             (push '("[X]"                  . "☑" ) prettify-symbols-alist)
                             (push '("lambda"               . "λ"  ) prettify-symbols-alist)
                             (push '("subtitle"             . "𝙩" ) prettify-symbols-alist)
                             (prettify-symbols-mode))))
#+end_src
**** Tag List
#+begin_src emacs-lisp

;;;;;; TAG LIST
(after! org

config.el
;;; $doomdir/config.el --- My Emacs Config File -*- lexical-binding: t; no-byte-compile: t; -*-
;;
;; author: marty buchaus <marty@dabuke.com>
;; copyright © 2021, marty buchaus, all rights reserved.
;; created:  1 November 2021
;;
;;;; Notes
;;
;;  2021 11 18 Update clean Install and config
;;  2021 10 12  added code from Stuff from  https://github.com/Artawower/.doom/blob/main/config.el#L308
;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;
;;; Code
;;;; GLOBAL

(setq user-full-name "Marty Buchaus")
(setq user-mail-address "marty@dabuke.com")
(setq epg-gpg-program "/usr/bin/gpg2")

(setq-default enable-local-variables t)            ; allow for reading the local variables file
(setq-default window-combination-resize t)
(setq-default left-margin-width 1)                 ; Define new widths
(setq-default right-margin-width 2)                ; Define new widths.
(setq-default x-stretch-cursor t)

(setq scroll-margin 2)                             ; it's nice to maintain a little margin
(setq evil-want-fine-undo t)                       ; by default while in insert all changes are one big blob. be more granular
(setq undo-limit 80000000)                         ; raise undo-limit to 80mb
(setq auto-save-default t)                         ; nobody likes to loose work, i certainly don't
(setq confirm-kill-emacs nil)                      ; stop hounding me and quit
(setq display-time-24hr-format t)                  ; I wonder what this does
(setq password-cache-expiry nil)                   ; i can trust my computers ... can't i?
(setq read-process-output-max (* 1024 1024))
(setq truncate-string-ellipsis "…")                ; unicode ellispis are nicer than "...", and also save /precious/ space
(setq warning-minimum-level :emergency)

(setq garbage-collection-messages nil)

(after! gcmh
  (setq gcmh-high-cons-threshold 67108864))  ; 33554432 32mb, or 67108864 64mb, or *maybe* 128mb, BUT NOT 512mb

(setq doom-scratch-initial-major-mode 'lisp-interaction-mode)  ; Make the scratch buffer start in lisp mode

(display-time-mode 1)                              ; enable time in the mode-line

(global-subword-mode 1)                            ; CamelCase and it makes refactoring slightly Essie

(after! projectile
  (setq projectile-indexing-method 'alien)
  (setq projectile-project-search-path '("~/Source")))
  ;; (setq projectile-project-root-files-bottom-up (remove ".git" projectile-project-root-files-bottom-up)))


(set-window-buffer nil (current-buffer))
(setenv "zstd" "/usr/bin/zstd")

;;;;; VTERM

(setq vterm-kill-buffer-on-exit t)
(setq vterm-always-compile-module t)               ; Always compile the vterm module
(setq vterm-shell "/usr/bin/zsh")

;;;; UI
;;;;; SET FONTS
(setq doom-font (font-spec :family "DejaVu Sans Mono" :size 15 :weight 'regular )
      doom-variable-pitch-font (font-spec :family "Ubuntu" :style "Regular" :size 15 :weight 'regular)
      doom-unicode-font (font-spec :family "symbola" :size 15)
      doom-big-font (font-spec :family "DejaVu Sans Mono" :size 24))

;;;;; FACES

(custom-set-faces!
  '(font-lock-comment-face :slant italic)
  '(font-lock-keyword-face :slant italic))
(setq global-prettify-symbols-mode t)

(custom-set-faces!
 '(mode-line :family "firacode nerd font mono" :height 100)
 '(mode-line-inactive :family "firacode nerd font mono" :height 100))

;;;;; THEME

(setq doom-theme 'doom-dracula )

(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))

;;;;; MODELINE

(after! doom-modeline
  (setq all-the-icons-scale-factor 1.1)
  (setq doom-modeline-enable-word-count t)
  (setq doom-themes-padded-modeline t)
  (setq auto-revert-check-vc-info t)
  (setq doom-modeline-buffer-file-name-style 'relative-to-project)
  (setq doom-modeline-major-mode-color-icon (display-graphic-p))
  (setq doom-modeline-major-mode-icon (display-graphic-p))
  (setq doom-modeline-vcs-max-length 60)

  (add-hook! 'doom-modeline-mode-hook
    (progn
      (set-face-attribute 'header-line nil
                          :background (face-background 'mode-line)
                          :foreground (face-foreground 'mode-line))
      )))

;;;;; LINE NUMBERS

(setq display-line-numbers-type 'relative)

;; remove numbers from these modes

(dolist (mode '(org-mode-hook
                term-mode-hook
                shell-mode-hook
                eshell-mode-hook))
  (add-hook mode (lambda () (display-line-numbers-mode 0))))

;;;; PACKAGES
;;;;; DIRED

(add-hook 'peep-dired-hook 'evil-normalize-keymaps)
;; get file icons in dired
(add-hook 'dired-mode-hook 'all-the-icons-dired-mode)
;; with dired-open plugin, you can launch external programs for certain extensions
;; for example, i set all .png files to open in 'sxiv' and all .mp4 files to open in 'mpv'
(setq dired-open-extensions '(("gif" . "sxiv")
                              ("jpg" . "sxiv")
                              ("png" . "sxiv")
                              ("mkv" . "mpv")
                              ("mp4" . "mpv")))
;;;;; COMPANY

(after! company
  (setq company-idle-delay 0.5)
  (setq company-backends
        '(company-capf company-dabbrev company-files company-yasnippet)
        company-global-modes '(not comint-mode erc-mode message-mode help-mode gud-mode)))

(use-package! company-box
  :after company
  :config
  (setq company-box-max-candidates 5))

(use-package! company-prescient
  :after company
  :hook (company-mode . company-prescient-mode))

;;;;; MAGIT

(after! magit
  (setq magit-commit-arguments '("--gpg-sign=7ACF5A8C6A106D1F561F1699E72D93984A1E2483")
        magit-rebase-arguments '("--autostash" "--gpg-sign=7ACF5A8C6A106D1F561F1699E72D93984A1E2483")
        magit-pull-arguments   '("--rebase" "--autostash" "--gpg-sign=7ACF5A8C6A106D1F561F1699E72D93984A1E2483"))
  (magit-define-popup-option 'magit-rebase-popup
    ?S "Sign using gpg" "--gpg-sign=" #'magit-read-gpg-secret-key)
  (setq magit-revision-show-gravatars '("^author:     " . "^commit:     ")))

;;;;; FILE-TEMPLATE


;;;; MODULES

(defun marty/autoinsert-yas-expand ()
  (let ((template ( buffer-string )))
    (delete-region (point-min) (point-max))
    (yas-expand-snippet template)
    (evil-insert-state)))

(use-package! autoinsert
  :init
  (setq auto-insert-query nil)
  (setq auto-insert-directory "~/.config/doom/templates")
  (add-hook 'find-file-hook 'auto-insert)
  (auto-insert-mode 1)
  :config
  (define-auto-insert "\\.html?$" "default.html")
  (define-auto-insert "\\.org" ["default.org" marty/autoinsert-yas-expand])
  (define-auto-insert "\\.sh" ["default.sh" marty/autoinsert-yas-expand])
  (define-auto-insert "\\.el" ["default.el" marty/autoinsert-yas-expand])
  (define-auto-insert "Roam/.+\\.org?$" ["defaultRoam.org" marty/autoinsert-yas-expand])
  (define-auto-insert "Blorg/snuffy-org/.+\\.org?$" ["snuffy-org.org" marty/autoinsert-yas-expand])
  (define-auto-insert "Sites/snuffy.org/.+\\.org?$" ["snuffy-org-posts.org" marty/autoinsert-yas-expand])
  (define-auto-insert "salt-master.+\\.org?$" ["salt-master.org" marty/autoinsert-yas-expand])
  (define-auto-insert "NSI-Documentation/[^/]+\\.org?$" ["NSI-Documentation.org" marty/autoinsert-yas-expand])
  (define-auto-insert "NSI-Documentation/.+/[^/]+\\.org?$" ["NSI-Documentation.org" marty/autoinsert-yas-expand])
  (define-auto-insert "NSI-Documentation/tipjar/[^/]+\\.org?$" ["NSI-Documentation-tipjar.org" marty/autoinsert-yas-expand])
  (define-auto-insert "NSI-Documentation/TVA/[^/]+\\.org?$" ["NSI-Documentation-TVA.org" marty/autoinsert-yas-expand])
  (define-auto-insert "NSI-Documentation/TVA/ScanReports/.+[^/]+\\.org?$" ["NSI-Documentation-TVA-scanreport.org" marty/autoinsert-yas-expand])
  (define-auto-insert "NSI-Documentation/Patching/.+[^/]+\\.org?$" ["NSI-Documentation-Patching-Notes.org" marty/autoinsert-yas-expand])
  (define-auto-insert "masons/[^/].+\\.org?$" ["masonsMeetingMinuets.org" marty/autoinsert-yas-expand])
  (define-auto-insert "daily/[^/].+\\.org?$" ["defaultRoamDaily.org" marty/autoinsert-yas-expand])
  (define-auto-insert "/[0-9]\\{8\\}.org$" ["defaultJournal.org" marty/autoinsert-yas-expand]))

;;;;; ACTIVITY WATCH MODE

(defun marty/startactivitywatchmode ()
  (interactive)
  (global-activity-watch-mode))

(use-package! activity-watch-mode
  :defer t
  :config
  (add-hook 'doom-first-buffer-hook #'marty/startactivitywatchmode))

;;;;; AGGRESSIVE INDENT

(use-package! aggressive-indent
  :config
  (add-hook 'emacs-lisp-mode-hook #'aggressive-indent-mode)
  (add-hook 'clojure-mode-hook    #'aggressive-indent-mode)
  (add-hook 'php-mode-hook #'aggressive-indent-mode)
  (add-hook 'hy-mode-hook #'aggressive-indent-mode))

;;;;; I3 WINDOW MANAGER CONFIG
;; Syntax highlighting for i3 config
(use-package! i3wm-config-mode
      :defer t )

;;;;; JENKINS

(use-package! jenkinsfile-mode
:defer t )

;;;;; KHARD

(use-package! khardel
  :defer t )

;;;;; NGINX

(use-package! company-nginx
  :after nginx-mode
  :config (add-hook 'nginx-mode-hook (lambda () (add-to-list 'company-backends #'company-nginx))))

(use-package! nginx-mode
  :defer t)

;;;;; MUTT-MODE

(use-package! mutt-mode
  :defer t)

;;;;; OUTSHINE

(use-package! outshine
  :defer t
  :hook (emacs-lisp-mode . outshine-mode)
  :config
  (map! :after outshine
        :map emacs-lisp-mode-map
        "TAB" #'outshine-cycle)
  (add-hook 'emacs-lisp-mode-hook #'outshine-mode)
  (defvar outline-minor-mode-prefix "\M-#"))

;;;;; PAPERLESS

(use-package! paperless
  :commands (paperless)
  :config (progn
            (custom-set-variables
             '(paperless-capture-directory "~/Nextcloud/Documents/INBOX")
             '(paperless-root-directory "~/Nextcloud/Documents")))
  (map! :after paperless
        :localleader
        :mode paperless-mode
        "d"  #'paperless-display
        "r"  #'paperless-rename
        "r"  #'paperless-scan-directories
        "f"  #'paperless-file
        "x"  #'paperless-execute))

;;;;; RAINBOW MODE

(use-package! rainbow-mode
  :hook (((css-mode scss-mode org-mode emacs-lisp-mode typescript-mode js-mode). rainbow-mode)))

;;;;; SALT MODE

(use-package! salt-mode
  :defer t)

;;;;; SYSTEMD MODE

(use-package! systemd
  :defer t
  :config
  (map! :map systemd-mode
        :localleader
        :prefix ("h" . "help")
        "d" #'systemd-doc-directives
        "o" #'systemd-doc-open))

;;;;; VLF

(use-package! vlf-setup
  :defer-incrementally  vlf-tune vlf-base vlf-write vlf-search vlf-occur vlf-follow vlf-ediff vlf)

;;;;; WAKATIME

(defun marty/startwakatime ()
  (interactive)
  (setq wakatime-api-key (auth-source-pass-get 'secret "Application/wakatime/apikey"))
  (global-wakatime-mode))

(use-package! wakatime-mode
  :defer t
  :config
  (add-hook 'doom-first-buffer-hook  #'marty/startwakatime)
  (setq wakatime-cli-path "/usr/bin/wakatime"))



;;;; LOAD
(load! "functions.el")
(load! "org-mode.el")
(load! "keybindings.el")
(load! "mu4e.el")
(load! "hydra.el")

;;; CUSTOM

(setq-default custom-file (expand-file-name ".custom.el" doom-private-dir))
(when (file-exists-p custom-file)
  (load custom-file))
